pinot-core/src/main/java/org/apache/pinot/server/access/AccessControl.java

```diff
@@ -27,6 +28,13 @@
 @InterfaceStability.Stable
 public interface AccessControl {
 
+  /**
+   *
+   * @param channelHandlerContext netty tls context
+   * @return Whether the client has access to query server
+   */
+  boolean hasQueryServerAccess(ChannelHandlerContext channelHandlerContext);
+
```

Comment:
If the desire is to achieve additional application-layer authentication, why not use an auth token here, similar to the rest of pinot?

Reply:
Yes. Both token and cert are used in Pinot to do authorization. For admin/query/segment dowload requests we check the tokens to ensure it's from authorized users. But in this case we only want server to accept tls connection from pinot-broker (the query requester's access is already checked), therefore, inspecting the host id on trusted certificate is sufficient. And it also incurs minimum overhead to check only upon handshake.

Reply:
I agree doing this on handshake is the right way of going about it.

For consistency in the existing auth model, a token-based approach seems like a better fit. We should be able to directly re-use the existing parsing mechanisms.

Reply:
The token based approach in existing auth model is extracting and parsing tokens from https request headers. For this server-broker channel, we are using netty tls channel and I'm not aware of approaches we can pass and inspect tokens in ChannelHandler.

Reply:
the next easiest thing would be passing an auth token on every request. This might be needed anyways for things like presto paging.

Reply:
We can pass auth token on every request, but sending and checking those tokens incurs overhead for especially high QPS cases.

Reply:
optimally, this would happen on establishing the channel. even per request the impact of a token check should be minimal. if it's too slow despite optimization, it could still be disabled.

Reply:
I'm not clear about the necessity to pass the token in the request given the potential overhead... For the other modules with token-based authorization, requests are all initiated by external services/users through exposed APIs, where we do need to check client tokens for access control. In this case the broker-server channel is private and we've already performed the token check on the broker side; servers can trust the broker's identity based on cert. Can you also give more details on why certificate-based check is not enough? Do we want to expose the servers' query endpoint?

Reply:
there are two scenarios here:
* broker - server
* presto - server (afaik this is exposed throught grpc right now)

If we implement application-layer authentication on the server, then we have to cover both cases. Right now, only transport layer security is supported via SSL certs.

The "cert-based check" mingles transport-layer and application-layer authentication. It also inverts the config dependency between servers and brokers, i.e. the servers have to know about the brokers' identities. This makes it tricky to handle dynamic reconfiguration, e.g. autoscaling. The alternative to checking "who I am" is to check "what I have", i.e. a token of some sort that can be pre-distributed.

Reply:
Is this something we can do as a follow-up ?

Reply:
sure. let's hold off on this interface change for now then. sounds like this needs a bit of planning

Reply:
I meant can "token based authZ" be done as a follow-up. We need broker -> server authZ. Is the ask here one of the following ?

- Do token based authZ instead of cert based. 
- Continue with "cert based authZ" like done in this PR but also do it for presto -> server communication.

Reply:
`Do token based authZ instead of cert based`. Ultimately, authz should work independently of the transport layer. the cert-based approach mingles the app and transport layer.

Reply:
We can't have token based authZ without SSL as it becomes less secure. Token has to be passed over an encrypted channel so using token based authZ will also be done as part of establishing the channel.

Reply:
First, establish a SSL connection, then pass a token to authenticate the channel. The advantages of using a layered approach are (a) authz can be tested separately from encryption, and (b) it's trivially portable to other transports like grpc for presto as well

Reply:
We have add an example of cert based access control `CertBasedTlsChannelAccessControl`  in `BasicAuthTlsRealtimeIntegrationTest.java`. We can add token based authZ as a follow-up, as will work independent from this cert-based solution.

Reply:
ok thank you